crates/polars-expr/src/expressions/mod.rs

```diff
@@ -662,3 +663,61 @@ pub trait PartitionedAggregation: Send + Sync + PhysicalExpr {
         state: &ExecutionState,
     ) -> PolarsResult<Series>;
 }
+
+pub struct AppendExpr {
+    left: Arc<dyn PhysicalExpr>,
+    right: Arc<dyn PhysicalExpr>,
+    upcast: bool,
+    expr: Expr,
+}
+
+impl AppendExpr {
+    pub fn new(
+        left: Arc<dyn PhysicalExpr>,
+        right: Arc<dyn PhysicalExpr>,
+        upcast: bool,
+        expr: Expr,
+    ) -> Self {
+        Self {
+            left,
+            right,
+            upcast,
+            expr,
+        }
+    }
+}
+
+impl PhysicalExpr for AppendExpr {
+    fn evaluate(&self, df: &DataFrame, state: &ExecutionState) -> PolarsResult<Series> {
+        let mut left_s = self.left.evaluate(df, state)?;
+        let mut right_s = self.right.evaluate(df, state)?;
+
+        if self.upcast {
+            let dt = try_get_supertype(left_s.dtype(), right_s.dtype())?;
+
+            left_s = left_s.cast(&dt)?;
+            right_s = right_s.cast(&dt)?;
```

Comment:
*question*: Should the casting be encoded in the IR, rather than plumbing `upcast` all the way through?

That is, if upcasting, should we have inserted symbolic casts in the expr->ir lowering phase?

Similarly, if not upcasting, can we complain early (during query lowering) if the dtypes don't match, rather than during execution?

Reply:
I would agree that this would be ideal, but I am just not so sure that it is feasible. We may not know the datatypes in the IR, so casting to a supertype with the datatypes there may not be reliable.

We could lower in the IR, but for it to be reliable it would require a `ExprIR::ToSharedSuperType` node.